src/coreclr/tools/Common/Compiler/AsyncContinuationType.cs

```diff
@@ -0,0 +1,85 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+using Internal.TypeSystem;
+
+namespace ILCompiler
+{
+    /// <summary>
+    /// An async continuation type. The code generator will request this to store local state
+    /// through an async suspension/resumption. We only identify these using a <see cref="GCPointerMap"/>
+    /// since that's all the code generator cares about - size of the type, and where the GC pointers are.
+    /// </summary>
+    public sealed partial class AsyncContinuationType : MetadataType
+    {
+        private readonly MetadataType _continuationBaseType;
+        public GCPointerMap PointerMap { get; }
+
+        public override DefType[] ExplicitlyImplementedInterfaces => [];
+        public override ReadOnlySpan<byte> Name => Encoding.UTF8.GetBytes(DiagnosticName);
+        public override ReadOnlySpan<byte> Namespace => [];
+
+        // We don't lay these out using MetadataType metadata.
+        // Autolayout (which we'd get due to GC pointers) would likely not match what codegen expects.
+        public override bool IsExplicitLayout => throw new NotImplementedException();
+        public override bool IsSequentialLayout => throw new NotImplementedException();
+        public override bool IsExtendedLayout => throw new NotImplementedException();
+        public override bool IsAutoLayout => throw new NotImplementedException();
+        public override ClassLayoutMetadata GetClassLayout() => throw new NotImplementedException();
+
+        public override bool IsBeforeFieldInit => false;
+        public override ModuleDesc Module => _continuationBaseType.Module;
+        public override MetadataType BaseType => _continuationBaseType;
+        public override bool IsSealed => true;
+        public override bool IsAbstract => false;
+        public override MetadataType ContainingType => null;
+        public override PInvokeStringFormat PInvokeStringFormat => default;
+        public override string DiagnosticName => $"ContinuationType_{PointerMap}";
```

Comment:
The name on the VM side is `Continuation_<data size>_<offset 0>_<num GCrefs 0>_<offset 1>_<num GCrefs 1>...`.
Should we make these match? (Not sure if it gets exposed to anything.)